src/sage/numerical/pivot_rules_for_simplex_method.py

```diff
@@ -0,0 +1,776 @@
+            x1
+            sage: D.leaving()
+            x4
+
+        Handling unboundedness, feasibility, and optimality should be left to the solver.::
+
+            sage: A = ([1, 0],)
+            sage: b = (1,)
+            sage: c = (0, 1)
+            sage: P = InteractiveLPProblemStandardForm(A, b, c)
+            sage: D = P.initial_dictionary()
+            sage: D.run_simplex_method()
+            \begin{equation*}
+            ...
+            \end{equation*}
+            The problem is unbounded in $x_{2}$ direction.
+
+        The leaving variable should be ``None`` if a dictionary represents unboundedness in a problem::
+
+            sage: D = P.initial_dictionary()
+            sage: BlandsRule.dictionary_pivot(D)
+            sage: D.entering()
+            x2
+            sage: D.leaving() is None
+            True
+
+        The :meth:`__call__` method of :class:`SimplexMethodPivot` is responsible checking for type input::
+
+            sage: LP_tablaux = Matrix([[0, 10, 5], [1, 1, 1000], [3, 1, 1500]])
+            sage: BlandsRule.dictionary_pivot(LP_tablaux)
+            Traceback (most recent call last):
+            ...
+            AttributeError: 'sage.matrix.matrix_integer_dense.Matrix_integer_dense' object has no attribute 'entering'
+
+            """
+    @abstract_method
+    def dual_dictionary_pivot(dictionary, *args):
+        r"""
+        Abstract method for computing a pivot for the dual simplex method based on an :class:`sage.numerical.interactive_simplex_method.LPDictionary`.
+
+        Sets the entering and leaving variables of :class:`sage.numerical.interactive_simplex_method.LPDictionary` for the dual simplex method.
+
+        One does not need to define a :meth:`dual_dictionary_pivot` unless intending to run a dual simplex algorithm.
+
+        INPUT:
+
+        - ``dictionary`` -- an instance of :class:`sage.numerical.interactive_simplex_method.LPDictionary`.
+
+        EXAMPLES::
+
+            sage: from sage.numerical.pivot_rules_for_simplex_method import BlandsRule
+            sage: A = ([1, 1], [3, 1], [-1, -1])
+            sage: b = (1000, 1500, -400)
+            sage: c = (10, 5)
+            sage: P = InteractiveLPProblemStandardForm(A, b, c)
+            sage: D = P.dictionary(2, 3, 5)
+            sage: D.entering()
+
+            sage: D.leaving()
+
+            sage: BlandsRule.dual_dictionary_pivot(D)
+            sage: D.entering()
+            x1
+            sage: D.leaving()
+            x3
+
+        Some pivot rules do not have dual methods implemented::
+
+            sage: from sage.numerical.pivot_rules_for_simplex_method import SteepestEdge
+            sage: SteepestEdge.has_dual_method()
+            False
+        """
+
+class SimplexMethodPivot(SageObject, UniqueRepresentation):
+    r"""
+    A class to perform pivots in the simplex method of solving Linear Programs.
+
+    INPUT:
+
+    - ``name`` -- a string or a ``SimplexMethodPivotRule`` object.
```

Comment:
always use Sphinx roles such as `` `:class:`SimplexMethodPivotRule` `` for references to classes, methods, etc